rust/headless-client/src/ipc_service.rs

```diff
@@ -492,13 +492,15 @@ impl<'a> Handler<'a> {
         )
         .map_err(|e| Error::PortalConnection(e.to_string()))?;
 
+        let dns = self.dns_controller.system_resolvers();
```

Comment:
```suggestion
        // Reading system_resolvers immediately after / during initializing the tun interface can result in an empty read, so read them here before.
        let dns = self.dns_controller.system_resolvers();
```

Reply:
Actually, did we prove that? Did we see a log where we sent an empty resolver list to connlib?

Reply:
Actually, yeah maybe that's not the root bug. I think this is it:

- We call set_dns before the interface is configured
- `update_dns_mapping` is effectively a no-op if the interface isn't configured, leaving DNS unset
- We never call set_dns **after** the tun interface is configured

Reply:
cc @thomaseizinger

Reply:
The code kinda looks like calling `set_dns` first just buffers the resolvers for later. Also in the PR I call `set_dns` first and that appears to be fine? Although we didn't test `ifconfig.net` on that build.

Maybe we should have written down our hypotheses while we were testing

Reply:
Hm yeah. But if we never call `set_dns` after the tun interface is configured, they will be left unset, connlib returns immediately and doesn't save them to its internal state.

See [here](https://github.com/firezone/firezone/pull/6455/files#diff-73548333208f7371e1de2b318331e296c7eb78c2b9ecc5c2edd68cc7411b3cb8R1125).

Reply:
Nah, the caller of that function does save it. I don't have a stack trace but I assume it flows through here: https://github.com/firezone/firezone/blob/a0ee0922ec55102bef7eca1cc1e5927a230815df/rust/connlib/tunnel/src/client.rs#L864-L868

Reply:
Ah, kk. Hmm maybe setting them in this order then just triggers a side effect combined with the path monitoring that results in empty resolvers somehow.